README.md

```diff
@@ -1,2 +1,89 @@
-# ibm-cloud-env-golang
-IBM Cloud Environment for Go
+# IBM Cloud Environment
+
+The `ibm-cloud-env-golang` module allows to abstract environment variables from various Cloud compute providers, such as, but not limited to, CloudFoundry and Kubernetes, so the application could be environment-agnostic.
+
+The module allows to define an array of search patterns that will be executed one by one until required value is found.
+
+### Installation
+
+```bash
+go get github.com/ibm-developer/ibm-cloud-env-golang
+```
+ 
+### Usage
+
+Create a JSON file containing your mappings and initialize the module
+
+```golang
+import "github.com/ibm-developer/ibm-cloud-env-golang"
+
+//in main function 
+IBMCloudEnv.Initialize("/path/to/the/mappings/file/relative/to/prject/root")
+```
+ 
+#### Supported search patterns types
+ibm-cloud-config supports searching for values using three search pattern types - cloudfoundry, env, file. 
+- Using `cloudfoundry` allows to search for values in VCAP_SERVICES and VCAP_APPLICATIONS environment variables
+- Using `env` allows to search for values in environment variables
+- Using `file` allows to search for values in text/json files
+
+#### Example search patterns
+- cloudfoundry:service-instance-name - searches through parsed VCAP_SERVICES environment variable and returns the `credentials` object of the matching service instance name
+- cloudfoundry:$.JSONPath - searches through parsed VCAP_SERVICES and VCAP_APPLICATION environment variables and returns the value that corresponds to JSONPath
+- env:env-var-name - returns environment variable named "env-var-name"
+- env:env-var-name:$.JSONPath - attempts to parse the environment variable "env-var-name" and return a value that corresponds to JSONPath
+- file:/server/config.text - returns content of /server/config.text file
+- file:/server/config.json:$.JSONPath - reads the content of /server/config.json file, tries to parse it, returns the value that corresponds to JSONPath
+- user-provided:service:name - searches through parsed VCAP_SERVICES environment variable and returns the value of the requested service name and credential
+
+#### mappings.json file example
+```javascript
+{
+    "service1-credentials": {
+        "searchPatterns": [
+            "cloudfoundry:my-service1-instance-name", 
+            "env:my-service1-credentials", 
+            "file:/localdev/my-service1-credentials.json" 
+        ]
+    },
+    "service2-username": {
+        "searchPatterns":[
+            "cloudfoundry:$.service2[@.name=='my-service2-instance-name'].credentials.username",
+            "env:my-service2-credentials:$.username",
+            "file:/localdev/my-service1-credentials.json:$.username",
+            "user-provided:my-service2-instance-name:username"
+
```

Comment:
In the above 2 sample search patterns, please add `user-provided` to both and put that entry in the first spot. The patterns are read in order and should go in this order:
		"searchPatterns": [
			"user-provided:conversation:url",
			"cloudfoundry:$['conversation'][0].credentials.url",
			"env:service_watson_conversation:$.url",
			"file:/server/localdev-config.json:$.watson_conversation_url"
		]

Reply:
We're not sure how to use user-provided for the first example. The first example grabs all the credentials, but user-provided can only grab a specific credential, not the entire thing